core/lib/Cardano/Address/KeyHash.hs

```diff
@@ -0,0 +1,313 @@
+        _ ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.cc_cold) $ keyHashAppendByteCIP0129 keyHash cred
+    CommitteeHot -> case govType of
+        Just CIP0105 ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.cc_hot_vkh) keyHash
+        _ ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.cc_hot) $ keyHashAppendByteCIP0129 keyHash cred
+    Unknown ->
+        T.decodeUtf8 $ encode EBase16 keyHash
+
+-- | In accordance to CIP-0129 (https://github.com/cardano-foundation/CIPs/tree/master/CIP-0129)
+--   one byte is prepended to vkh only in governance context. The rules how to contruct it are summarized
+--   below
+--
+--   drep       0010....
+--   hot        0000....    key type
+--   cold       0001....
+--
+--   keyhash    ....0010
+--   This is on top of X_vkh, where X={drep, cc_hot, cc_hot}, which lacks the additional byte.
+--   In `keyHashFromText` we additionally
+--   support reading legacy X which also lacks the additional byte, and has the same payload as
+--   as the corresponding X_vkh.
+keyHashAppendByteCIP0129 :: ByteString -> KeyRole -> ByteString
+keyHashAppendByteCIP0129 payload cred =
+    maybe payload (`BS.cons` payload) bytePrefix
+  where
+    bytePrefix = case cred of
+        Representative -> Just 0b00100010
+        CommitteeCold -> Just 0b00010010
+        CommitteeHot -> Just 0b00000010
+        _ -> Nothing
+
+-- | Construct a 'KeyHash' from 'Text'. It should be
+-- Bech32 encoded text with one of following hrp:
+-- - `addr_shared_vkh`
+-- - `stake_shared_vkh`
+-- - `addr_vkh`
+-- - `stake_vkh`
+-- - `policy_vkh`
+-- - `drep`
+-- - `cc_cold`
+-- - `cc_hot`
+-- - `drep_vkh`
+-- - `cc_cold_vkh`
+-- - `cc_hot_vkh`
+-- - `addr_shared_vk`
+-- - `stake_shared_vk`
+-- - `addr_vk`
+-- - `stake_vk`
+-- - `policy_vk`
+-- - `cc_cold_vk`
+-- - `cc_hot_vk`
+-- - `addr_shared_xvk`
+-- - `stake_shared_xvk`
+-- - `addr_xvk`
+-- - `stake_xvk`
+-- - `policy_xvk`
+-- - `drep_xvk`
+-- - `cc_cold_xvk`
+-- - `cc_hot_xvk`
+-- Raw keys will be hashed on the fly, whereas hash that are directly
+-- provided will remain as such.
+-- If if hex is encountered Unknown policy key is assumed
+--
+-- @since 3.1.0
+keyHashFromText :: Text -> Either ErrKeyHashFromText KeyHash
+keyHashFromText txt =
+    case (fromBase16 $ T.encodeUtf8 txt) of
+        Right bs ->
+            if checkBSLength bs 28 then
+                pure $ KeyHash Unknown bs
+            else if checkBSLength bs 32 then
+                pure $ KeyHash Unknown (hashCredential bs)
+            else if checkBSLength bs 64 then
+                pure $ KeyHash Unknown (hashCredential $ BS.take 32 bs)
+            else
+                Left ErrKeyHashFromTextInvalidHex
+        Left _ -> do
+            (hrp, dp) <- first (const ErrKeyHashFromTextInvalidString) $
+                Bech32.decodeLenient txt
+
+            maybeToRight ErrKeyHashFromTextWrongDataPart (Bech32.dataPartToBytes dp)
+                >>= maybeToRight ErrKeyHashFromTextWrongHrp . convertBytes hrp
```

Comment:
If `convertBytes` returned an `Either` you wouldn't need `maybeToRight ErrKeyHashFromTextWrongHrp`

Reply:
yes, but `convertBytes` was one of the first functions added in this project and is used all around. I would not touch. But I agree with your diagnosis